File: LondonFhirService.Core/Services/Foundations/ResourceMatchers/PracticeQualifiedDdsMatchKey.cs
// ---------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------

using System.Text.Json;

namespace LondonFhirService.Core.Services.Foundations.ResourceMatchers
{
    /// <summary>
    /// The DDS identifier qualified by the practice that issued it.
    ///
    /// A DDS identifier is unique within one practice's feed, not within a bundle. A patient
    /// registered at more than one practice gets a record built from both, and both number their
    /// resources from the same sequence - so a record spanning two practices carries two
    /// unrelated resources both claiming DDS id '54'.
    ///
    /// Keyed on the bare identifier those collide, and the matcher does the safe thing with a
    /// collision: keeps the first and sets the rest aside as manual-review-required. Safe, but
    /// wrong - and the reason it writes reads "not found in Source2" when both sides hold the
    /// resource perfectly well. Worth being clear about what the alternative would have been: had
    /// it matched them, it would have compared one practice's sublingual temperature against the
    /// other's tympanic one and reported confident, meaningless differences.
    ///
    /// Qualifying with the ODS code off meta.tag separates them, so each practice's '54' pairs
    /// with its own counterpart.
    ///
    /// A resource with no ODS tag keys on an empty prefix rather than being dropped. That leaves
    /// untagged resources grouped exactly as they were before this existed - no better, no worse -
    /// rather than inventing a distinction the data does not carry.
    /// </summary>
    internal static class PracticeQualifiedDdsMatchKey
    {
        private const string DdsIdentifierSystem = "https://fhir.hl7.org.uk/Id/dds";
        private const string OdsCodeSystem = "https://fhir.nhs.uk/Id/ODS-Code";

        public static string Build(JsonElement resource)
        {
            string ddsIdentifier = GetDdsIdentifier(resource);

            // No DDS identifier means no key at all, which is what every one of these matchers
            // did before and what the portal mirrors when it decides a resource has no row of
            // its own to hang a difference on.
            if (string.IsNullOrWhiteSpace(ddsIdentifier))
            {
                return null;
            }

            return $"{GetOdsCode(resource)}|{ddsIdentifier}";
        }

        private static string GetDdsIdentifier(JsonElement resource)
        {
            if (!resource.TryGetProperty("identifier", out JsonElement identifiers)
                || identifiers.ValueKind != JsonValueKind.Array)
            {
                return null;
            }

            foreach (JsonElement identifier in identifiers.EnumerateArray())
            {
                if (identifier.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                if (identifier.TryGetProperty("system", out JsonElement system)
                    && system.ValueKind == JsonValueKind.String
                    && system.GetString() == DdsIdentifierSystem
                    && identifier.TryGetProperty("value", out JsonElement value)
                    && value.ValueKind == JsonValueKind.String)
                {
                    return value.GetString();
                }
            }

            return null;
        }

        private static string GetOdsCode(JsonElement resource)
        {
            if (!resource.TryGetProperty("meta", out JsonElement meta)
                || meta.ValueKind != JsonValueKind.Object)
            {
                return string.Empty;
            }

            if (!meta.TryGetProperty("tag", out JsonElement tags)
                || tags.ValueKind != JsonValueKind.Array)
            {
                return string.Empty;
            }

            foreach (JsonElement tag in tags.EnumerateArray())
            {
                if (tag.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                if (tag.TryGetProperty("system", out JsonElement system)
                    && system.ValueKind == JsonValueKind.String
                    && system.GetString() == OdsCodeSystem
                    && tag.TryGetProperty("code", out JsonElement code)
                    && code.ValueKind == JsonValueKind.String)
                {
                    return code.GetString() ?? string.Empty;
                }
            }

            return string.Empty;
        }
    }
}

File: LondonFhirService.Manage.Client/src/helpers/comparisons/practiceQualifiedDdsMatchKey.ts
import { findIdentifierBySystem, readObjectArray, readPath, readString } from "../fhir/fhirJson";
import type { FhirResource } from "../../models/foundations/fhir/FhirResource";

const ddsIdentifierSystem = "https://fhir.hl7.org.uk/Id/dds";
const odsCodeSystem = "https://fhir.nhs.uk/Id/ODS-Code";

// Mirrors PracticeQualifiedDdsMatchKey server side, which every matcher that pairs on a DDS
// identifier uses - Observation, FamilyMemberHistory, Immunization, Encounter, MedicationRequest,
// DiagnosticReport, Procedure, ProcedureRequest, ReferralRequest and Appointment - so a difference
// the engine recorded against one of these resources can be shown on that resource's own row.
//
// A DDS identifier is unique within one practice's feed, not within a bundle: a patient
// registered at two practices carries two unrelated resources both claiming DDS id '54'. Keying
// on the bare identifier collided, and the engine reported those collisions as
// manual-review-required rather than pairing them. The ODS code off meta.tag separates them.
//
// A resource with no ODS tag keys on an empty prefix rather than returning null - the same as the
// server - so untagged resources group exactly as they did before this existed.
//
// This has to stay in step with the server: the key is how a difference finds the row it belongs
// to, so a change to one side without the other leaves differences unattached and they fall into
// the Other differences list instead.
export function buildPracticeQualifiedDdsMatchKey(resource: FhirResource): string | null {
    const ddsIdentifier = readString(findIdentifierBySystem(resource, ddsIdentifierSystem)?.value);

    if (ddsIdentifier === null) {
        return null;
    }

    return `${readOdsCode(resource) ?? ""}|${ddsIdentifier}`;
}

function readOdsCode(resource: FhirResource): string | null {
    const tags = readObjectArray(readPath(resource, "meta", "tag"));

    for (const tag of tags) {
        if (readString(tag.system) === odsCodeSystem) {
            return readString(tag.code);
        }
    }

    return null;
}
